microcloud/cmd/microcloud/main_init.go

```diff
@@ -262,154 +262,112 @@ func lookupPeers(s *service.Handler, autoSetup bool, subnet *net.IPNet, systems
 	return nil
 }
 
-func AddPeers(sh *service.Handler, systems map[string]InitSystem) error {
-	joinConfig := make(map[string]types.ServicesPut, len(systems))
-	secrets := make(map[string]string, len(systems))
-	for peer, info := range systems {
-		if peer == sh.Name {
-			continue
-		}
-
-		joinConfig[peer] = types.ServicesPut{
-			Tokens:     []types.ServiceToken{},
-			Address:    info.ServerInfo.Address,
-			LXDConfig:  info.JoinConfig,
-			CephConfig: info.MicroCephDisks,
-		}
-
-		secrets[peer] = info.ServerInfo.AuthSecret
-	}
-
+// waitForJoin issues a token and instructs a system to request a join,
+// and then waits for the request to either complete or time out.
+// If the request was successful, it additionally waits until the cluster appears in the database.
+func waitForJoin(sh *service.Handler, clusterSize int, secret string, peer string, cfg types.ServicesPut) error {
 	mut := sync.Mutex{}
 	err := sh.RunConcurrent(false, false, func(s service.Service) error {
-		for peer := range joinConfig {
-			token, err := s.IssueToken(peer)
-			if err != nil {
-				return fmt.Errorf("Failed to issue %s token for peer %q: %w", s.Type(), peer, err)
-			}
-
-			mut.Lock()
-			cfg := joinConfig[peer]
-			cfg.Tokens = append(joinConfig[peer].Tokens, types.ServiceToken{Service: s.Type(), JoinToken: token})
-			joinConfig[peer] = cfg
-			mut.Unlock()
+		token, err := s.IssueToken(peer)
+		if err != nil {
+			return fmt.Errorf("Failed to issue %s token for peer %q: %w", s.Type(), peer, err)
 		}
 
+		mut.Lock()
+		cfg.Tokens = append(cfg.Tokens, types.ServiceToken{Service: s.Type(), JoinToken: token})
+		mut.Unlock()
+
 		return nil
 	})
 	if err != nil {
 		return err
 	}
 
-	fmt.Println("Awaiting cluster formation ...")
-	// Initially add just 2 peers for each dqlite service to handle issues with role reshuffling while another
-	// node is joining the cluster.
-	initialSize := 2
-	if len(joinConfig) > initialSize {
-		initialCfg := map[string]types.ServicesPut{}
-		for peer, info := range joinConfig {
-			initialCfg[peer] = info
-
-			if len(initialCfg) == initialSize {
-				break
-			}
+	cloud := sh.Services[types.MicroCloud].(*service.CloudService)
+	errChan := cloud.RequestJoin(context.Background(), secret, peer, cfg)
```

Comment:
Rather than implementing this timeout, can we pass a context with timeout to RequestJoin?